### PR TITLE
Dynamic response sender count

### PR DESCRIPTION
**Pull request by andy31415**
#### Problem
Large platforms that have heap have no reason to limit minmdns number of operational network advertisers.

#### Change overview
Add a config option for limiting mdns advertisers:
  - fixed number will use an array just as before (except using std::array for easier type usage)
  - dynamic count (setting number to 0) will use a std::list and not limit anything.

#### Testing
Compile and unit tests. 
Devkit-C commissioning.
CI will validate some stress testing.

### CODE REVIEW

File: src/lib/core/CHIPConfig.h [hunk 1]
```diff
@@ -1771,6 +1771,22 @@ extern const char CHIP_NON_PRODUCTION_MARKER[];
...
+ *        advertise in parallel.
+ *
+ *        Value will be used for internal CHIP MDNS implementations but may
+ *        be ignored for platform implementations where the platform implements
+ *        its own memory management for DNSSD advertisement.
+ *
+ *        Set to 0 to mean "unlimited/use heap".
+ */
+#ifndef CHIP_CONFIG_MAX_OPERATIONAL_NETWORKS
+#define CHIP_CONFIG_MAX_OPERATIONAL_NETWORKS 5
```
Comment (Damian-Nordic) [inline src/lib/core/CHIPConfig.h:1787]: How is that different from `CHIP_CONFIG_MAX_FABRICS`?

Comment (andy31415) [reply src/lib/core/CHIPConfig.h:1787]: Well beside the obvious "I did not know about MAX_FABRICS"  this does support 0 for using heap for operational network advertisement.

However I also agree this is duplicating (and numbers seem to differ ... I was told by spec this number should be 5 however I see a lot of MAX_FABRICS being set to 4 on platforms) so maybe just a "USE_HEAP" would be a more correct flag. Looking into it.

File: src/lib/core/core.gni [hunk 2]
```diff
@@ -53,6 +53,18 @@ declare_args() {
 
   # When enabled traces messages using pw_trace.
   chip_enable_transport_trace = false
+
+  # Defines the maximum number of operational networks that the device can
+  # support advertising if using internam dnssd implementations.
```
[the remark this answers is not in the document]
Comment (andy31415) [reply src/lib/core/core.gni:58]: Typo for `internal`. Changed it to 'minmdns' however now based on the other feedback I think this should be just a bool for using heap instead. Will rework the CL in that direction.

File: src/lib/core/CHIPConfig.h [hunk 3]
```diff
@@ -1771,6 +1771,20 @@ extern const char CHIP_NON_PRODUCTION_MARKER[];
...
 
+/*
+ * @def CHIP_CONFIG_MINMDNS_DYNAMIC_OPERATIONAL_RESPONDER_LIST
+ *
+ * @brief Enables usage of heap in the minmdns DNSSD implementation
+ *        for tracking active operational responder lists.
+ *
+ *        When this is not set, CHIP_CONFIG_MAX_FABRICS is used to determine
+ *        and statically allocate pointers needed to track active
+ *        operational responder lits.
```
Comment (bzbarsky-apple) [inline src/lib/core/CHIPConfig.h:1782]:
```suggestion
 *        operational responder lists.
```

File: src/lib/dnssd/minimal_mdns/ResponseSender.h [hunk 4]
```diff
@@ -88,8 +105,13 @@ class ResponseSendingState
 class ResponseSender : public ResponderDelegate
 {
 public:
-    // TODO(cecille): Template this and set appropriately. Please see issue #8000.
+    // Generally this is 2 + number of operational networks required, based on
+    // the current implementation of Advertiser_ImplMinimalMdns.cpp:
+    //    - 1 for commissionable advertising
+    //    - 1 for commissioner responder
+    //    - extra for every operational advertisement
     static constexpr size_t kMaxQueryResponders = 7;
```
Comment (bzbarsky-apple) [inline src/lib/dnssd/minimal_mdns/ResponseSender.h:113]: Should this also be using CHIP_CONFIG_MAX_FABRICS?  That said, this seems to be test-only now.  Should this move into the test?

**Review by bzbarsky-apple — APPROVED** [review]

File: src/lib/dnssd/minimal_mdns/ResponseSender.h (hunk 4 above)
Comment (andy31415) [reply src/lib/dnssd/minimal_mdns/ResponseSender.h:113]: updated code and test (test now loops in an add and checks that only added once).

**Comment by woody-apple** [comment]
Fast tracking, given this has had enough time for review.

**Pull request merged**